lib/model/autocomplete.dart

```diff
@@ -78,6 +105,30 @@ final RegExp mentionAutocompleteMarkerRegex = (() {
     unicode: true);
 })();
 
+final RegExp _emojiIntentRegex = (() {
+  // Similar reasoning as in _mentionIntentRegex.
+  // Specifically forbid a preceding ":", though, to make "::" not a query.
+  const before = r'(?<=^|\s|\p{Punctuation})(?<!:)';
+  // TODO(dart-future): Regexps in ES 2024 have a /v aka unicodeSets flag;
+  //   if Dart matches that, we could combine into one character class
+  //   meaning "whitespace and punctuation, except not `:`":
+  //     r'(?<=^|[[\s\p{Punctuation}]--[:]])'
+
+  /// Characters that might be meant as part of (a query for) an emoji's name,
+  /// other than whitespace.
+  const nameCharacters = r'_\p{Letter}\p{Number}';
```

Comment:
Hmm, how do we know which characters these are? For the corresponding part in `_mentionIntentRegex` we exclude characters, not include them.

Here, if we don't include `\p{Emoji}` and I guess maybe the ZWJ (maybe more?) then we defeat the part of `EmojiAutocompleteQuery.matches` that's supposed to enable searching by literal emoji.

I don't think that feature is really useful here actually. If I want a specific emoji from the iOS emoji picker, I'll just open that directly, right, instead of typing ":" first. Searching by literal emoji seems more helpful when you're choosing an emoji reaction than when you're typing an emoji into a message.

I guess I don't mind offering the feature. 🤷‍♂️ It might help someone who wants their platform's emoji picker but has already typed ":" because they weren't sure if they needed to.

Reply:
Yeah, I guess this means searching by literal emoji isn't actually offered in the message-content autocomplete. Which is fine because as you say, it's not that useful here.

What that feature is really for is for adding an emoji reaction — it effectively gives the user the option of using their phone keyboard's emoji picker instead of ours.



> Hmm, how do we know which characters these should be? For the corresponding part in `_mentionIntentRegex` we exclude characters, not include them.

Yeah, which matches the fact that users' names can be almost anything, with a small list of character exclusions. Emoji names are more constrained.

But I hadn't actually looked up what those exact constraints are. Turns out it's more constrained even than I'd guessed:
```py
def check_valid_emoji_name(emoji_name: str) -> None:
    if emoji_name:
        if re.match(r"^[0-9a-z\-_]+(?<![\-_])$", emoji_name):
            return
    # [… otherwise raise one error or other.]